salt/config/__init__.py

```diff
@@ -3360,13 +3422,18 @@ def get_id(opts, cache_minion_id=False):
         log.debug('Guessing ID. The id can be explicitly set in {0}'
                   .format(os.path.join(salt.syspaths.CONFIG_DIR, 'minion')))
 
-    newid = salt.utils.network.generate_minion_id()
+    if opts.get('id_function'):
+        newid = eval_id_function(opts)
+    else:
+        newid = salt.utils.network.generate_minion_id()
```

Comment:
I fixed it so that if the function returns an empty string, it will hard fail instead of failover to FQDN. I believe this is the preferred behavior because a user is using the `id_function` feature specifically because they want to avoid the default FQDN behavior. As a user of this feature, that is definitely the behavior I would prefer because I don't want FQDN IDs getting to the master, since this is not what is expected and would cause problems based on how my master is set up.